suit: Initial minimal CBOR-based SUIT manifest parser

bergzand opened the pull request:
### Contribution description

Minimal version of the CBOR-based SUIT parser, provides minimal supports for [version 1 of the draft specification](https://tools.ietf.org/html/draft-moran-suit-manifest-01).
Supported retrieval functions are:
 - manifest version
 - timestamp/sequence number
 - conditions
 - url
 - payloadinfo:
    - digest algorithm
    - digest
    - storage ID
    - size

### Testing procedure

Unittest included :)

### Issues/PRs references

#### Todo:
 - [x] split uuid generation

bergzand commented:
> Would you be able to update it to the latest SUIT version (v3, [here](https://datatracker.ietf.org/doc/draft-moran-suit-manifest/))? I'll review it before v4 for sure.

After some offline online discussion with @emmanuelsearch we decided to keep this the v1.

> Not tested: suit_uuid_xxx functions (apart from init); suit_isnew. Could you please include explicit tests?

Will do.

> This is a _CBOR_ SUIT parser specifically; other serialization formats are possible. Could you please indicate this in the module name OR abstract away the CBOR-specific stuff so that we can use other serialization formats with this code?

Version 3 of the draft explicitly calls is a "CBOR-based Firmware Manifest Serialisation Format". Despite this would you still like to have support for different serialization formats?

> Maybe take the reference to block1 out of the brief, because we have block 2 now as well (i.e. so it just says you need an in-place buffer)

Done

> Please split uuid_init as planned (although keeping this function to initialise everything together would also be useful)

On it.

bergzand commented:
I've reworked the suit parser to:

1. Split of the conditionals and UUID handling. A separate module to handle more flow related actions will be in a follow up PR.
2. The `suit_manifest_t` is reduced to only a pointer and length to the stored SUIT manifest. It is parsed on demand for the necessary information. This increases the processing burden, but reduced the required memory and flash a bit.

kaspar030 commented on sys/include/suit.h:
```diff
@@ -0,0 +1,225 @@
+#define SUIT_H
+
+#ifdef __cplusplus
+extern "C" {
```
pls move this below the includes

bergzand replied:
Out of interested (don't want to spark a discussion on this), is there a rule/guideline for the order of this somewhere?

bergzand replied:
Changed btw

kaspar030 replied:
> is there a rule/guideline for the order of this somewhere?

"do as we always did" combined with "close in reversed open order":

- copyright
- doxygen open (with file info)
- header guard open
- includes
- ifdef c++ open
- code
- ifdef c++ close
- header guard close
- doxygen close

for the includes it is somewhat important to be before the "extern C" as they *might* have other namespacing.

bergzand replied:
Thanks for the clarification

kaspar030 commented on sys/include/suit.h:
```diff
@@ -0,0 +1,225 @@
+#ifdef __cplusplus
+}
+#endif
+/** @} */
```
maybe move this line below the next - so it is

```
/** @{ */
#header guard open
...
#header guard close
/* @} */
```

(if you have the file open anyways)

bergzand replied:
Done

kaspar030 commented on sys/suit/suit.c:
```diff
@@ -0,0 +1,326 @@
+    return 0;
+}
+
+static int parse_manifest_version(CborValue *it,
```
please rename ```parse_*``` to ```_parse_*``` if static.

bergzand replied:
I should know better, changed 5 occurences.

kaspar030 commented on sys/suit/suit.c:
```diff
@@ -0,0 +1,326 @@
+
+static int _advance_x(CborValue *it, unsigned x)
+{
+    for(unsigned i=0; i < x; i++)
```
uncrustify

bergzand replied:
Uncrustified!

kaspar030 commented on sys/suit/suit.c:
```diff
@@ -0,0 +1,326 @@
+    }
+    return SUIT_OK;
+}
+
```
extra newline

bergzand replied:
Removed

danpetry commented:
Thanks for the nudge @kaspar030 . To comment on the above:
> After some offline online discussion with @emmanuelsearch we decided to keep this the v1.
Just interested, what's the reasoning? We keep v1 until the non-draft RFC comes out, rather than updating several times? This seems reasonable to me.
> Version 3 of the draft explicitly calls is a "CBOR-based Firmware Manifest Serialisation Format". Despite this would you still like to have support for different serialization formats?

Pagel is also working on a binary serialization format and corresponding RFC:
https://datatracker.ietf.org/wg/suit/documents/
On the mailing list there has been quite a lot of discussion about several potential formats. It's my understanding that there will come other formats. (Correct me if I'm wrong.) This is the main one atm though. What about keeping as is but calling the module suit_cbor? We can make it modular later when we wish to support other formats.

I also think it is cleaner to have this contribution focusing on getting the information from the manifest, with update application logic submitted later, as discussed offline.

Once these points are addressed I'm happy with the high level aspects.

emmanuelsearch commented:
@danpetry 

> We keep v1 until the non-draft RFC comes out, rather than updating several times? 

Yup, updating later is the plan, indeed. Depending on how long the standardisation process will last, we'll consider updating to draft version X>3.

> What about keeping as is but calling the module suit_cbor? 

+1 for something like this, makes sense to me.

bergzand commented:
> What about keeping as is but calling the module suit_cbor? We can make it modular later when we wish to support other formats.

You've convinced me, will do.

bergzand commented:
@danpetry @emmanuelsearch renamed ~everything to `suit_cbor`

kaspar030 commented:
@bergzand would you mind rebasing this? Let's see what Murdock says. I suspect it'll want some defines to be documented.

bergzand commented:
> @bergzand would you mind rebasing this?

Rebased!

>  I suspect it'll want some defines to be documented.

Not sure what you're talking about :innocent:

danpetry commented:
Section 3.7 in v1 of the [draft manifest](https://tools.ietf.org/pdf/draft-moran-suit-manifest-01.pdf) gives a list of minimum features that MUST be supported by the target. It looks like, out of this list, these aren't implemented here:

4. timestamp
 5. conditions

Is that the case? Strongly think that if we're saying this is adherent to the draft (as indicated in the header) then these should be implemented.

danpetry commented:
Sorry I missed this before.

danpetry commented:
>  @danpetry @emmanuelsearch renamed ~everything to `suit_cbor`

Thanks :)

emmanuelsearch commented:
thanks for your vigilance @danpetry 

In  practice, nodes with small micro-controllers such as those targeted by RIOT have no notion of the absolute time on-board.
Two options to solve this issue:
1. assume some upstream computer (e.g. the software update repo) serves manifests only if the timestamp check passes;
2. assume the node can poll a trusted remote server which can indicate the current time;

Both options are out of scope for this PR. 

For the conditionals, I let @bergzand answer ;)

That said: 
- I agree we should indicate in the comments something like "the implementation complies with v1 specification, except BLAH and BLAH to be addressed later";
- this is only a *parser* and not the full process of validation, so upcoming complementary PRs are expected anyhow.

The idea here is to get preliminary building blocks in master, to PR before the IETF104 Hackathon a ready-to-use example of RIOT firmware update using a suit-cbor metadata manifest.
Then at the hackathon, different people could split work to collaborate on next steps (support for more features, upgrade to v4 etc.)

Makes sense?

danpetry commented:
This makes sense to me for sure. In particular that we can use the opportunity of the IETF to push merging of a full main branch OTA update application example.

However I think we should still be careful of merging partial implementations of modules, which to the observer might appear as complete implementations before they discover later that they are only partial. (They might think, "which other specs are only partially implemented which we don't know about?")

Do we consider a note in the API reference to be enough to stop users from making this mistake?

waehlisch commented:
Agree with @danpetry that we should be very careful with announcing what we support. In this case, we cannot claim to provide a "Minimal version of the SUIT parser". I think a clear warning in the API documentation is the very minimum. But you are right @danpetry, the note in the API might be insufficient as the directory name of the implementation gives a different impression.

Anyhow, I don't understand why we are in a rush. For the hackathon, you can provide a ready-to-use example based on a branch that is not master.

Regarding the topic: how to handle cases where the target does not provide absolute time on-board: This should be discussed in the SUIT WG, and we should implement the outcome of this discussion. The result of the discssion might require updating the draft (e.g., in case of option 1).

bergzand commented:
@danpetry Thanks for the suggestions.

> timestamp

If I'm correct with my interpretation of the draft, this is supported in this PR, the `suit_cbor_get_seq_no` provides the sequence number of the manifest.

> conditions

Conditionals are intentionally not supported yet. My plan is to include this in a follow up PR, but as it is now, I have to think a bit on how to structure the api to make them accessible in a convenient way for an actual SUIT manifest updater.

As stated before by @emmanuelsearch this PR only provides a very minimal parser for SUIT CBOR manifests. There is no handling of the information contained in the manifest. I don't mind adding some comments on what is supported at the moment, but IMHO, the API reference should be enough to give developers an idea of what we support.

emmanuelsearch commented:
> this should be discussed in the SUIT WG

+1 yes, that is part of the plan. I am preparing an email to the mailing list as we speak.

> For the hackathon, you can provide a ready-to-use example based on a branch that is not SUITmaster.

@waehlisch at the SUIT workshop last June we also "had a branch". It's not necessarily the best approach, because it quickly becomes a jungle few can navigate, and it results in more time to extract elements from that which actually end in master.
Going purposely in smaller steps merging bits quicker in master (assuming they *work*) was the idea this time around.
We do not want to herald *WE HAVE SUIT SUPPORT* just yet, even when this PR gets merged -- the spec is work-in-progress anyways.
We want to put ourselves in the best position to have something more people can test and contribute to.

> what other alerts could we include?
@danpetry Feel free to propose some complementary warnings ;)
let's just remember it's just a parser, so let's keep things into perspective?

waehlisch commented:
> Going purposely in smaller steps merging bits quicker in master (assuming they _work_) was the idea this time around.

Is only reasonable

1. if there is long-term commitment, i.e., updating the implementation along to the evolvement of the preliminary specs. I see this long-term commitment given.
1. the description of the contribution is correct, i.e., does describe the current state of implementation. This is not given here as this PR does not implement a "Minimal version of the SUIT parser". Btw, "minimal" is an ambiguous term in this context.
1. the same rigorousness of code review is applied as for any other code in master.

waehlisch commented:
> @emmanuelsearch 

> If I'm correct with my interpretation of the draft, this is supported in this PR, the `suit_cbor_get_seq_no` provides the sequence number of the manifest.

@bergzand and why do you don't call it `suit_cbor_get_timestamp`?

bergzand commented:
> @bergzand and why do you don't call it `suit_cbor_get_timestamp`?

Because (and I don't have a different source than "from conversations with @bremoran" for this now) the "timestamp" is supposed to be a strictly monotonic sequence number which *could* be a timestamp, but is not supposed to be used as a timestamp. In v2 of the draft it is called "sequence". I decided to implement it here as `suit_cbor_get_seq_no` to prevent developers from considering it a timestamp and to prevent confusion with the best before conditional which *is* using a timestamp.

bremoran commented:
@bergzand You're correct. Here's a reference: https://tools.ietf.org/html/draft-ietf-suit-information-model-02#section-3.2

emmanuelsearch commented:
@bremoran so in a nutshell
- a timestamp + notion of time on-board is optional, 
- a monotonically increasing sequence number is a MUST.
In this case, we are fine here ;)

bremoran commented:
@emmanuelsearch You've got it! This has caused no end of confusion. We (Milosch and I) originally recommended using UTC timestamps as monotonic sequence numbers because they guarantee world-wide monotonic increase of sequence numbers. This seemed useful in coordinating disparate development teams. However, the use of timestamps in any way causes so much confusion on embedded systems. This eventually meant that we had to reword things significantly, calling it a sequence number and suggesting that a good source of monotonic sequence numbers is UTC time.

waehlisch commented:
Implementing version x and using the wording of version y introduces confusion as well ;).

danpetry commented:
Can I propose moving forward with the following:
1. An alert to the developer at build time, informing them that the spec is subject to change and they should check the API documentation so they know what is supported
2. Leaving `seq_no` in place of `timestamp` as there are stronger engineering arguments for `seq_no` and point 1. should avoid confusion
3. Using the forthcoming SUIT-based updater implementation to drive requirements for the handling of conditionals

bergzand commented:
> Implementing version x and using the wording of version y introduces confusion as well ;).

> Leaving `seq_no` in place of `timestamp` as there are stronger engineering arguments for `seq_no` and point 1. should avoid confusion

With this being the cause of some confusion, I can add a bit of clarification to the function documentation (and if preferred also to the top level documentation) to clarify the `timestamp`/`seq_no` naming.

bergzand commented:
> An alert to the developer at build time, informing them that the spec is subject to change and they should check the API documentation so they know what is supported

Do we have a preferred mechanism for this in the build system? (@cladmi?)

bergzand commented:
> Using the forthcoming SUIT-based updater implementation to drive requirements for the handling of conditionals

My plan is to write the glue code between RIOT and the library today/tomorrow. Would you prefer to have this PR updated with the conditionals getter or have them in a follow-up? I have no preference for one or the other, so whatever makes the reviewers life easier :-)

kaspar030 commented:
> Do we have a preferred mechanism for this in the build system?

Not do that at all. We're talking about a small part (manifest parser) of an implementation for a draft of a standard. Just add known missing functionality (that are required by the draft) with a small note to the docs.

bergzand commented:
@danpetry I've added conditions support to the parser, functions are available to retrieve the condition ID and condition parameter. Can you confirm that the parser now supports the minimum required feature set?

@waehlisch I've added a small note to clarify that the `suit_cbor_seq_no` function retrieves what the draft calls the timestamp.

kaspar030 commented:
I think squashing is OK at this point.

bergzand commented:
@kaspar030 Squashed into two commits, one for the parser and one for the unittest

bergzand commented:
Amended an uncrustify for the unittest in

danpetry commented:
ACK on high level aspects - I'm happy with the approach and testing but haven't looked at the code and internal design in detail.

The only difference I can see between CBOR SUIT v1 and this is the nomenclature of the timestamp/seq no. Reasonably arguments have been given for the use of the term "sequence number". API documentation on the state of the PR seems good to me - people can be reasonably expected to read the documentation and this is now a minor difference. If there are further thoughts about what kind of things should elicit a build system warning, IMO this should be continued in a separate channel.

As discussed offline, this will be updated to more recent versions of the SUIT draft in subsequent PRs.

emmanuelsearch commented:
tested on native and samr21

emmanuelsearch commented:
Looks like we're good for this PR. Looking forward to the next ones ;)
@kaspar030 I leave you the honor of hitting the green button!

kaspar030 commented:
@waehlisch have your issues been resolved?

waehlisch commented:
@bergzand can you please update the description of this PR? "Minimal version of the SUIT parser, more to follow" is not helpful as discussed before.

Otherwise fine. Looking forward to quick alignement with more recent spec versions.

bergzand commented:
> @bergzand can you please update the description of this PR? "Minimal version of the SUIT parser, more to follow" is not helpful as discussed before.

Updated to something more descriptive, thanks for the heads-up

kaspar030 commented:
> Updated to something more descriptive, thanks for the heads-up

@waehlisch please hit merge when satisfied!

miri64 commented on sys/include/suit/cbor.h:
```diff
@@ -0,0 +1,273 @@
+ * The parser is based on draft version 1 of the specification, restricted to
+ * handling CBOR based manifests.
+ *
+ * @see https://tools.ietf.org/html/draft-moran-suit-manifest-01
```
Maybe add the `@experimental` tag as well, since this is based on a draft?

miri64 replied:
see

https://github.com/RIOT-OS/RIOT/blob/e288c9b23fcbfc6744ca377aa3a27cbb0ad19d9f/doc/doxygen/riot.doxyfile#L231

bergzand replied:
Done

bergzand commented:
I've added the `@experimental` tag as suggested by @miri64. I've amended directly, see the [force push](https://github.com/RIOT-OS/RIOT/compare/48330463c3544a928e44a69c145b0b8829b127d6..ea15dd221dea715f7b13260726a046dcc18ffd5e) diff for verification.

kaspar030 commented:
@waehlisch please hit merge when satisfied!

bergzand commented:
> un-ACK, v4 is fundamentally different, let's see if we still need this.

Agreed!

emmanuelsearch commented:
@bergzand closing this one since we are now planning to get SUIT support through the cluster of PRs around #11818 (but do feel free to reopen if you think it is necessary!)
